src/StreamUtils.bsv

```diff
@@ -0,0 +1,367 @@
+import Vector::*;
+import FIFOF::*;
+import GetPut::*;
+import SemiFifo::*;
+
+typedef 8 BYTE_WIDTH;
+typedef TLog#(BYTE_WIDTH) BYTE_WIDTH_WIDTH;
+typedef TMul#(4, BYTE_WIDTH) DWORD_WIDTH;
+
+typedef 2 CONCAT_STREAM_NUM;
+
+typedef 512 DATA_WIDTH;
+typedef TDiv#(DATA_WIDTH, BYTE_WIDTH) BYTE_EN_WIDTH;
+typedef 'hFFFFFFFFFFFFFFFF MAX_BYTE_EN;
+
+typedef Bit#(DATA_WIDTH) Data;
+typedef Bit#(BYTE_EN_WIDTH) ByteEn;
+typedef Bit#(TAdd#(1, TLog#(DATA_WIDTH))) DataBitPtr;
+typedef Bit#(TAdd#(1, TLog#(BYTE_EN_WIDTH))) DataBytePtr;
+
+typedef 32 STREAM_SIZE_WIDTH;
+typedef UInt#(STREAM_SIZE_WIDTH) StreamSize;
+typedef Bit#(TAdd#(1, TLog#(STREAM_SIZE_WIDTH))) StreamSizeBitPtr;
+
+typedef struct {
+    Data data;
+    ByteEn byteEn;
+    Bool isFirst;
+    Bool isLast;
+} DataStream deriving(Bits, Bounded, Eq, FShow);
+
+typedef struct {
+    DataStream stream;
+    DataBytePtr bytePtr;
+} StreamWithPtr deriving(Bits, Bounded, Eq, FShow);
+
+interface StreamConcat;
+    interface FifoIn#(DataStream)  inputStreamFirstFifoIn;
+    interface FifoIn#(DataStream)  inputStreamSecondFifoIn;
+    interface FifoOut#(DataStream) outputStreamFifoOut;
+endinterface
+
+interface StreamSplit;
+    interface FifoIn#(DataStream)  inputStreamFifoIn;
+    interface FifoIn#(StreamSize)  splitLocationFifoIn;
+    interface FifoOut#(DataStream) outputStreamFifoOut;
+endinterface
+
+function DataBytePtr convertByteEn2BytePtr (ByteEn byteEn);
+    ByteEn byteEnTemp = byteEn;
+    DataBytePtr ptr = 0;
+    while (byteEnTemp > 0) begin
+        byteEnTemp = byteEnTemp >> 1;
+        ptr = ptr + 1;
+    end
+    return ptr;
+endfunction
+
+function DataStream getEmptyStream ();
+    return DataStream{
+        data: 0,
+        byteEn: 0,
+        isFirst: False,
+        isLast: True
+    };
+endfunction
+
+function DataBitPtr getMaxBitPtr ();
+    return fromInteger(valueOf(DATA_WIDTH));
+endfunction
+
+function DataBytePtr getMaxBytePtr ();
+    return fromInteger(valueOf(BYTE_EN_WIDTH));
+endfunction
+
+// Concat two DataStream frames into one. StreamA.isLast must be True, otherwise the function will return a empty frame to end the stream.
+function Tuple3#(DataStream, DataStream, DataBytePtr) getConcatStream (DataStream streamA, DataStream streamB, DataBytePtr bytePtrA, DataBytePtr bytePtrB);
+    Bool isCallLegally = (streamA.isLast && bytePtrA <= getMaxBytePtr() && bytePtrA > 0);
+    DataBitPtr bitPtrA = zeroExtend(bytePtrA) << fromInteger(valueOf(BYTE_WIDTH_WIDTH));
+
+    // Fill the low PtrA bytes by streamA data
+    Data concatDataA = streamA.data;
+    ByteEn concatByteEnA = streamA.byteEn;
+
+    // Fill the high bytes by streamB data
+    Data concatDataB = streamB.data << bitPtrA;
```

Comment:
Please align them to make the code pretty. I know there are a lot of code to change, but if you make it a habbit, then you will write beautiful code in the future.
```    
    Data     concatDataB     = streamB.data << bitPtrA;
    ByteEn   concatByteEnB   = streamB.byteEn << bytePtrA;
    Data     concatData      = concatDataA | concatDataB;
    ByteEn   concatByteEn    = concatByteEnA | concatByteEnB;
```